pkg/otlp/model/translator/statspayload.go

```diff
@@ -0,0 +1,471 @@
+	TopLevelHits uint64
+}
+
+// aggregations stores aggregation values (stats) grouped by their corresponding keys.
+type aggregations struct {
+	agg map[aggregationKey]*aggregationValue
+}
+
+// Value returns the aggregation value corresponding to the key found in map m.
+func (a *aggregations) Value(m pcommon.Map) *aggregationValue {
+	var sntx bool
+	if v, ok := m.Get(statsKeySynthetics); ok {
+		sntx = v.Bool()
+	}
+	key := aggregationKey{
+		Service:        getStr(m, statsKeyService),
+		Name:           getStr(m, statsKeySpanName),
+		Resource:       getStr(m, statsKeySpanResource),
+		HTTPStatusCode: uint32(getInt(m, statsKeyHTTPStatusCode)),
+		Type:           getStr(m, statsKeySpanType),
+		DBType:         getStr(m, statsKeySpanDBType),
+		Synthetics:     sntx,
+	}
+	if a.agg == nil {
+		a.agg = make(map[aggregationKey]*aggregationValue)
+	}
+	if _, ok := a.agg[key]; !ok {
+		a.agg[key] = new(aggregationValue)
+	}
+	return a.agg[key]
+}
+
+// Stats returns the set of pb.ClientGroupedStats based on all the aggregated key/value
+// pairs.
+func (a *aggregations) Stats() []pb.ClientGroupedStats {
+	cgs := make([]pb.ClientGroupedStats, 0, len(a.agg))
+	for k, v := range a.agg {
+		cgs = append(cgs, pb.ClientGroupedStats{
+			Service:        k.Service,
+			Name:           k.Name,
+			Resource:       k.Resource,
+			HTTPStatusCode: k.HTTPStatusCode,
+			Type:           k.Type,
+			DBType:         k.DBType,
+			Synthetics:     k.Synthetics,
+			Hits:           v.Hits,
+			Errors:         v.Errors,
+			Duration:       v.Duration,
+			OkSummary:      v.OkSummary,
+			ErrorSummary:   v.ErrorSummary,
+			TopLevelHits:   v.TopLevelHits,
+		})
+	}
+	return cgs
+}
+
+// UnsetHostnamePlaceholder is the string used as a hostname when the hostname can not be extracted from span attributes
+// by the processor. Upon decoding the metrics, the Translator will use its configured fallback SourceProvider to replace
+// it with the correct hostname.
+//
+// This isn't the most ideal approach to the problem, but provides the better user experience by avoiding the need to
+// duplicate the "exporter::datadog::hostname" configuration field as "processor::datadog::hostname". The hostname can
+// also not be left empty in case of failure to obtain it, because empty has special meaning. An empty hostname means
+// that we are in a Lambda environment. Thus, we must use a placeholder.
+const UnsetHostnamePlaceholder = "__unset__"
+
+// StatsPayloadToMetrics converts OTLP Metrics to an APM Stats Payload.
+func (t *Translator) StatsPayloadFromMetrics(mx pmetric.Metrics) (pb.StatsPayload, error) {
+	var stats pb.StatsPayload
+	rmxs := mx.ResourceMetrics()
+	for i := 0; i < rmxs.Len(); i++ {
+		rmx := rmxs.At(i)
+		attr := rmx.Resource().Attributes()
+		if v, ok := attr.Get(keyAPMStats); !ok || !v.Bool() {
+			t.logger.Debug(fmt.Sprintf("StatsPayload Translator: was asked to convert metrics to stats payload, but identifier key %q was not present. Skipping.", keyAPMStats))
```

Comment:
Don't use formatted strings with the zap logger, use fields instead for performance

Reply:
That would put the field at the end as JSON, making it a pretty ugly log message. I could inline the text, or define the message globally, but what performance are we talking about here? This is generally called once every 10 seconds, and this log line should in reality never happen.

Reply:
> That would put the field at the end as JSON, making it a pretty ugly log message

Formatting will depend on where this is used and how you have configured the logger, in the Agent this would not be JSON, in the Collector it will be by default but you can configure it to have it look as you wish (and it's very common there to have fields on logging messages)

The performance impact will be small in this case, to me this is more of a 'this is not how zap's authors recommend using this API' and 'setting a good example in our codebase' than anything else.

Reply:
I agree with what you said, and you may have noticed I've used fields wherever it made sense, but here I think it doesn't. This is the current message:
```
was asked to convert metrics to stats payload, but identifier key "_dd.apm_stats" was not present. Skipping.
```
If I were to use a field, it would be:
```
was asked to convert metrics to stats payload, but identifier key was not present. Skipping {"key": "_dd.apm_stats"}
```
The latter seems very unusual to me. Why would you do that? I see fields as context to the message, am I wrong? I don't think that's the case here.